### PR TITLE
Enable dynamism on fill_

### PR DESCRIPTION
**Pull request by vanbasten23**
- added the code. Need to make the test pass next.
- add some couts

### CODE REVIEW

**Comment by vanbasten23** [comment]
I print out the xla_shape at the beginning of `fill_` https://github.com/pytorch/xla/blob/b248e21f78e86fb2246f6d3555202a090fe82701/torch_xla/csrc/tensor_methods.cpp#L1184-L1188. It appears to me functionalization would make in-place op lose dynamic shape info.

With functionalization enabled, I see result
```
root@t1v-n-2e71f932-w-0:/workspaces/work# TORCH_SHOW_DISPATCH_TRACE=1 save_ir   XLA_EXPERIMENTAL="nonzero:masked_select"    python3 pytorch/xla/test/ds/test_dynamic_shapes.py TestDynamicShapes.test_fill_  2>&1 | tee out.txt
xw32, file=/workspaces/work/pytorch/xla/torch_xla/csrc/tensor_methods.cpp, line=1191function=fill_: input->shape()=s32[6,1]{1,0}, input->shape().dynamic_dimensions()=0,0
```
Without functionalization, I see result:
```
root@t1v-n-2e71f932-w-0:/workspaces/work# XLA_DISABLE_FUNCTIONALIZATION=1   XLA_EXPERIMENTAL="nonzero:masked_select"    python3 pytorch/xla/test/ds/test_dynamic_shapes.py TestDynamicShapes.test_fill_  2>&1 | tee out.txt
xw32, file=/workspaces/work/pytorch/xla/torch_xla/csrc/tensor_methods.cpp, line=1191function=fill_: input->shape()=s32[<=6,1]{1,0}, input->shape().dynamic_dimensions()=1,0
```

**Comment by vanbasten23** [comment]
Running with `TORCH_SHOW_DISPATCH_TRACE=1` outputs:

```
root@t1v-n-2e71f932-w-0:/workspaces/work# TORCH_SHOW_DISPATCH_TRACE=1   XLA_EXPERIMENTAL="nonzero:masked_select"    python3 pytorch/xla/test/ds/test_dynamic_shapes.py TestDynamicShapes.test_fill_  2>&1 | tee out.txt

 [call] op=[aten::empty.memory_format], key=[PythonDispatcher]
  [redispatch] op=[aten::empty.memory_format], key=[CPU]
 [call] op=[aten::to.device], key=[PythonDispatcher]
  [call] op=[aten::_to_copy], key=[PythonDispatcher]
   [redispatch] op=[aten::_to_copy], key=[PythonDispatcher]

    [call] op=[aten::to.dtype_layout], key=[PythonDispatcher]
     [redispatch] op=[aten::to.dtype_layout], key=[PythonDispatcher]
 [call] op=[aten::lift_fresh], key=[PythonDispatcher]
 [call] op=[aten::detach_], key=[PythonDispatcher]
 [call] op=[aten::nonzero], key=[PythonDispatcher]
  [redispatch] op=[aten::nonzero], key=[PythonDispatcher]
   [callBoxed] op=[aten::nonzero], key=[PythonDispatcher]

 [call] op=[aten::fill_.Scalar], key=[PythonDispatcher]
  [redispatch] op=[aten::fill_.Scalar], key=[PythonDispatcher]
   [redispatch] op=[aten::fill_.Scalar], key=[PythonDispatcher]

    [call] op=[aten::empty.memory_format], key=[PythonDispatcher]
     [redispatch] op=[aten::empty.memory_format], key=[CPU]
    [call] op=[aten::item], key=[PythonDispatcher]
     [call] op=[aten::_local_scalar_dense], key=[PythonDispatcher]
    [call] op=[aten::fill_.Scalar], key=[PythonDispatcher]
    [call] op=[aten::fill.Scalar], key=[PythonDispatcher]
     [call] op=[aten::empty_like], key=[PythonDispatcher]
      [call] op=[aten::empty.memory_format], key=[PythonDispatcher]
       [redispatch] op=[aten::empty.memory_format], key=[XLA]
```

**Comment by vanbasten23** [comment]
I was able to find that when I do `t2.fill_(1)` in python with functionalization enabled, `XLATensor` are created twice :

1. [callstack](https://gist.github.com/vanbasten23/09dd107c7b3f3ea12f2854ea0212d372). We materialize the SizeNode. This is because during functionalization, we check `check_size_nonnegative`.
 
2. [callstack](https://gist.github.com/vanbasten23/d181676f7f24fbbec59504d113207fc5), `XLANativeFunctions::empty_symint(c10::ArrayRef<c10::SymInt> sym_size` is called. Note, the argument `sym_size` doesn't contain any dynamic dimensions.

**Comment by vanbasten23** [comment]
I got the [IR dump](https://gist.github.com/vanbasten23/07301a4a4297792645d87358146bdba4) of the test.  Looking at the IR dump
```
IR {
  %0 = s32[] prim::Constant(), xla_shape=s32[], value=1
  %1 = s32[6,1]{1,0} aten::expand(%0), xla_shape=s32[6,1]{1,0}, size=(6, 1), dynamic_dims=(0, 0), ROOT=0
}
```
I guess we lost the dynamism when we materialize the size node. But how come `aten::expand(%0)` return a shape `s32[6,1]{1,0}` when `%0` has shape `s32[]`.

**Comment by JackCaoG** [comment]
```
  %1 = s32[6,1]{1,0} aten::expand(%0), xla_shape=s32[6,1]{1,0}, size=(6, 1), dynamic_dims=(0, 0), ROOT=0
```
means, expand `%0`, to shape `s32[6,1]` and `dynamic_dim` is `(0,0)` means all static. You want to look at where `expand` is being called.

**Comment by vanbasten23** [comment]
I found the [callstack](https://gist.github.com/vanbasten23/1d22ed0ea26ad7c00caeb30e254a348d) when expand is called: it's called by `torch_xla::XLANativeFunctions::fill_`/`torch_xla::tensor_methods::fill_(XLATensorPtr& input,`.  Notice that I printed out `input->shape()` which gave me `input->shape()=s32[6,1]{1,0}, input->shape().dynamic_dimensions()=0,0`.

**Comment by miladm** [comment]
@bdhirsh wdyt about the input issue raised in this [comment](https://github.com/pytorch/xla/pull/4876#issuecomment-1507734369)?

**Comment by bdhirsh** [comment]
Hmm, so staring at the second call-stack, it looks like:

(1) functionalization turns `x.fill_(val) into `at::fill(x, val)`
(2) XLA doesn't implement `fill()`, so it relies on a `CompositeExplicitAutograd` implementation, `at::native::fill(x, val)`
(3) `fill(x, val)` calls `torch.empty_like(x)` to call create the new tensor, which calls `torch.empty(x.sizes())` [here](https://github.com/pytorch/pytorch/blob/6c5fdde881c24329b4356e085a7b171cfd68bf72/aten/src/ATen/native/TensorFactories.cpp#L394).

What's surprising is that if `x` has dynamic shapes, then calling `x.sizes()` should fail. So `x` (from `x.fill_(..)`) seems like it already has static shapes by the time it enters the functionalization kernel.

Can you confirm that? You could either use gdb, or manually add some prints. From the stack trace, you can see that the functionalization kernel you want to check is at `pytorch/build/aten/src/ATen/RegisterFunctionalization_2.cpp` (it's a codegen'd file). You can try printing `x.sizes()` and `x.sym_sizes()`, around line 5048 (the beginning of the functionalization kernel for `fill__Scalar`).

**Comment by vanbasten23** [comment]
Thanks @bdhirsh 

I printed out at the beginning of `at::Tensor & fill__Scalar(c10::DispatchKeySet dispatchKeySet, at::Tensor & self, const at::Scalar & value)`:
```
(gdb) p (self.sym_sizes()[0]).toSymNodeImpl()->str()
$7 = "<=6"
```
So it seems the tensor `self` still has the dynamic dimension.

**Comment by vanbasten23** [comment]
It makes me wonder if `nonzero` returns an incorrect result. So I did some testing and it turns out `nonzero` returns the correct result:

```
at::Tensor XLANativeFunctions::nonzero(const at::Tensor& self) {
  ...
  at::Tensor ret = bridge::AtenFromXlaTensor(tensor_methods::nonzero(self_tensor));
  auto sym_sizes = ret.sym_sizes();
  auto intArrOpt = c10::asIntArrayRefSlowOpt(sym_sizes);
  std::cout << "xw32, file=" << __FILE__ << ", line=" << __LINE__ << "function=" << __FUNCTION__ << ": c10::asIntArrayRefSlowOpt(ret.sym_sizes()).has_value()=" << intArrOpt.has_value() << std::endl;

  return ret;
}
```
I found:
```
(gdb) p intArrOpt.has_value()
[New Thread 0x7ff78c5cc700 (LWP 1459792)]
$3 = false
```

**Comment by vanbasten23** [comment]
Hi @ezyang, do you know when c10::SymInt.data_ changes? Specifically, if we have a c10::SymInt representing a dynamic size, then we materialize the size to get its real size, how should c10::SymInt.data_ get changed? Should it remain the same or change to the real size? Currently, it changed to the upper bound, so could it be a bug?

Here is more context about the issue. We are enabling dynamism when functionalization is enabled on `torch.fill_` and the test below
```
t1 = torch.tensor([1, 0, 3, 5, 0, 6], device=dev)
t2 = torch.nonzero(t1)
t2.fill_(1)
print('t2.shape=', t2.shape)
print('t2=', t2)
```
prints
```
t2.shape= torch.Size([6, 1])
t2= tensor([[1],
        [1],
        [1],
        [1],
        [1],
        [1]], device='xla:0', dtype=torch.int32)
```
Clearly the shape is wrong. We expect `t2.shape= torch.Size([4, 1])`.
What we found is that the python program `t2.fill_(1)` internally does 2 things in c++:
1. We materialize the SizeNode ([callstack](https://gist.github.com/vanbasten23/8d87f63545aa7c689bfb6b343f3bae93))
2. The functionalization kernel is called `at::Tensor & fill__Scalar(c10::DispatchKeySet dispatchKeySet, at::Tensor & self, const at::Scalar & value)`, which internally called `XLANativeFunctions::empty_symint(c10::ArrayRef<c10::SymInt> sym_size` ([callstack](https://gist.github.com/vanbasten23/d181676f7f24fbbec59504d113207fc5)).

The bug/issue happens in step2 above: the first dimension of `at::Tensor & self` lost dynamism:
```
(gdb) p (self.sym_sizes()[0]).toSymNodeImpl()->str()
$7 = "<=6"
```
which seems `self.sym_sizes()[0]` is dynamic, but
```
(gdb) p c10::asIntArrayRefSlowOpt(self.sym_sizes()).has_value()
$18 = true
```
which suggests none of the dimensions of tensor `self` is dynamic. These two outputs seems contradictory. So I print out the `self.sym_sizes()[0].data_` inside `/workspaces/work/pytorch/c10/core/SymInt.h`
```
216│   static bool check_range(int64_t i) {
217├>    return i > MAX_UNREPRESENTABLE_INT;
218│   }
``` 
which gives me:
```
(gdb) p i
$18 = 6
```
Also I print out the SymInt:
```
(gdb) p sym_size[0]
$27 = (const c10::SymInt &) @0x55555b15b440: {static MASK = 16140901064495857664, static IS_SYM = 11529215046068469760, static MAX_UNREPRESENTABLE_INT = -4611686018427387905, data_ = 6}
```
As a comparison, in the earlier step, before the code exits `torch_xla::XLANativeFunctions::nonzero`, I printed the same:
```
(gdb) p sym_sizes[0]
$25 = (const c10::SymInt &) @0x55555b15daf0: {static MASK = 16140901064495857664, static IS_SYM = 11529215046068469760, static MAX_UNREPRESENTABLE_INT = -4611686018427387905, data_ = -6917435202552494096}

(gdb) p i
$26 = -6917435202552494096
```

**Comment by ezyang** [comment]
We never change `data_`. Instead, the thing it points to has to change from being symbolic to a constant. This also leads to an annoying situation where you can have a SymInt that happens to be constant, but c'est la vie.

**Comment by ezyang** [comment]
Yes? This seems obviously true to me, so maybe I misunderstood the question.

**Comment by vanbasten23** [comment]
> Yes? This seems obviously true to me, so maybe I misunderstood the question.

Per my previous [comment](https://github.com/pytorch/xla/pull/4876#issuecomment-1518415730), what we observed today is that after the change the `constant` becomes the upper bound instead of the real size. So it looks like a bug in PyTorch.

**Comment by ezyang** [comment]
We don't have this problem in core PyTorch, so I guess it's some XLA thing. Do you ever return the upper bound in a situation where a concrete int is requested? That could cause this problem.

**Comment by vanbasten23** [comment]
hi @JackCaoG , I followed your suggestion to print out `XLATensor.GetUniqueId()` in the XLATensor constructor but I'm not sure how that would help. During `torch.fill_`, XLATensor constructor have been called twice and each outputs a different unique id.

My question now is when we create the XLATensor during `torch.fill_` ([callstack](https://gist.github.com/vanbasten23/d181676f7f24fbbec59504d113207fc5), how did we create the `at::Tensor & self` in `at::Tensor & fill__Scalar(c10::DispatchKeySet dispatchKeySet, at::Tensor & self, const at::Scalar & value)` (inside pytorch/build/aten/src/ATen/RegisterFunctionalization_2.cpp). I wonder if you have any tips on that.

**Comment by ezyang** [comment]
Functionalization will convert the `fill_` call into a functional `fill` call which is responsible for doing the ultimate allocation

**Comment by vanbasten23** [comment]
> Functionalization will convert the fill_ call into a functional fill call which is responsible for doing the ultimate allocation

I found something interesting here. During `torch.fill_` under the context of functionalization ([callstack](https://gist.github.com/vanbasten23/d181676f7f24fbbec59504d113207fc5)), `at::native::empty_like` in `/workspaces/work/pytorch/aten/src/ATen/native/TensorFactories.cpp` is called:
```
Tensor empty_like(...) {
 392│     } else {
 393│       // See Note [Explicit nullopt MemoryFormat argument]
 394│       // xw32 here
 395│       c10::optional<at::IntArrayRef> int_sizes = c10::asIntArrayRefSlowOpt(self.sym_sizes());
 396│       bool all_dims_static = int_sizes.has_value();
 397├>      result = at::empty(self.sizes(), options.memory_format(self.suggest_memory_format()), c10::nullopt);
 398│     }
 399│   } else {
 400│     // See Note [Explicit nullopt MemoryFormat argument]
 401│     result = at::empty(self.sizes(), options.memory_format(memory_format), c10::nullopt);
 402│   }
}
```
I added line395&line396 locally. I found that `all_dims_static` at line 396 is false which suggests one of the dimension of tensor `self` is dynamic. But at line397, `self.sizes()` is used instead of `self.sym_sizes()`. It doesn't seems right, does it?:
1. At line 397, should we use `self.sym_sizes()` instead of `self.sizes()`?
2. If the tensor `self` has a dynamic dimension, why does `self.sizes()` not throw an exception?

**Comment by vanbasten23** [comment]
I agree there seem to be 2 issues.

1. After I convert those `.sizes()` calls to `.sym_sizes()`, I got the correct results. Today, dynamic shape in Pytorch/XLA has not been integrated with Dynamo so I think the issue is irrelevant to Dynamo or AOTAutograd. But our tests do use `pd = torch._C._EnablePythonDispatcher()` which suggests CompositeAutogradImplicit is used. So it sounds torch_xla need to change https://github.com/pytorch/pytorch/blob/d7fa7fa8cfa086b9358c929cbf02e24f8cbe01a1/aten/src/ATen/native/TensorFactories.cpp#L398 to
```
result = at::empty_symint(self.sym_sizes(), options.memory_format(self.suggest_memory_format()), c10::nullopt);
```
?

2. Regarding "calling `t.sizes()` should raise an error if `t` has symbolic sizes", I had the same impression previously because a while I ago I implemented it in this [PR](https://github.com/pytorch/xla/pull/4553). You are correct that XLATensorImpl overrides `.sizes_custom()` but [`has_symbolic_sizes_strides_`](https://github.com/pytorch/xla/blob/24c792d6c6f377c3c6c7a02f001a9ef9b1bcf93f/torch_xla/csrc/tensor_impl.cpp#L117) is false somehow and that's why `t.sizes()` didn't raise an error. Please correct me if I'm wrong, `has_symbolic_sizes_strides_` seem to be set in `at::navtive::empty_cpu`:
```
 5122│ at::Tensor wrapper_CPU_memory_format_empty(c10::SymIntArrayRef size, c10::optional<at::ScalarType> dtype, c10::optional<at::Layout
 5123│     // No device check
 5124│   // DeviceGuard omitted
 5125├>  return at::native::empty_cpu(C10_AS_INTARRAYREF_SLOW(size), dtype, layout, device, pin_memory, memory_format);
 5126│ }
``` 
in /workspaces/work/pytorch/build/aten/src/ATen/RegisterCPU.cpp. Do we need to check if `c10::SymIntArrayRef size` has dynamic dimensions before converting it to an int array via `C10_AS_INTARRAYREF_SLOW`?

**Comment by ezyang** [comment]
What is probably happening is that in PT2 we have decomps that get applied when we run make_fx which will handle symbolic shapes correctly. However, XLA is *only* using Python dispatcher, and not the other decomps, so you end up exercising CompositeExplicitAutograd kernels here which we haven't symintified. One reasonable way to handle this is to symintify these kernels, so I guess I will take these patches. The other way is to patch XLA to start using our other decomps, note that this will increase trace time overhead though.

**Comment by ezyang** [comment]
> Do we need to check if c10::SymIntArrayRef size has dynamic dimensions before converting it to an int array via C10_AS_INTARRAYREF_SLOW?

I'm not entirely sure what the real Q is, but the answer to this specific question is that C10_AS_INTARRAYREF_SLOW will assert that all the integers are in fact constant before doing the conversion.

**Comment by vanbasten23** [comment]
hi @ezyang @bdhirsh , per our meeting, I'll hold this PR until pytorch symintifies the `empty_like` in /workspaces/work/pytorch/aten/src/ATen/native/TensorFactories.cpp (if my understanding is correct). Please let me know if there is anything I can help. Thanks.

**Comment by ezyang** [comment]
Oh, perhaps it wasn't clear in the meeting, but we were hoping y'all could submit the PR to symintify it and cc us on it.  This is esp because we don't actually use the empty_like c++ codepath with dynamic shapes, so non-xla wouldn't actually benefit from it.

**Comment by vanbasten23** [comment]
> Oh, perhaps it wasn't clear in the meeting, but we were hoping y'all could submit the PR to symintify it and cc us on it. This is esp because we don't actually use the empty_like c++ codepath with dynamic shapes, so non-xla wouldn't actually benefit from it.

SG. Let me send a PR your way.

File: test/pjrt/test_experimental_pjrt_tpu.py [hunk 1]
```diff
@@ -70,8 +70,12 @@ def test_xla_devices_multiprocess(self):
       'This test is not currently supported on v2 TPUVMs or earlier.')
   def test_xla_devices_single_process_all_chips(self):
     accelerator_devices = {
-        'v3-8': {i: torch.device(f'xla:{i}') for i in range(8)},
-        'v4-8': {i: torch.device(f'xla:{i}') for i in range(4)},
+        'v3-8': {
```
Comment (JackCaoG) [inline test/pjrt/test_experimental_pjrt_tpu.py:73]: does yapf changed this?

File: test/pytorch_test_base.py [hunk 2]
```diff
@@ -604,8 +604,8 @@ def skipped_test(self, *args, reason=reason, **kwargs):
             setattr(cls, dtype_test_name, disallowed_test)
           if not skipped:
             xla_dtypes.append(
-                dtype_combination
-                if len(dtype_combination) > 1 else dtype_combination[0])
+                dtype_combination if len(dtype_combination) >
```
Comment (JackCaoG) [inline test/pytorch_test_base.py:607]: ditto

File: torch_xla/csrc/tensor_methods.cpp [hunk 3]
```diff
@@ -1183,8 +1183,10 @@ void eye_out(XLATensorPtr& out, int64_t lines, int64_t cols) {
 }
 
 void fill_(XLATensorPtr& input, const at::Scalar& value) {
+  SymIntElements sym_int_elements(input->GetIrValue(), input->shape());
```
Comment (JackCaoG) [inline torch_xla/csrc/tensor_methods.cpp:1186]: can you add a comment here to explain why we need `sym_int_elements` here? Something along the line of 
```
Fill_ is implemented by expanding the pass in value to the same shape of input and replace input Tensor's IR with the expanded value. In order to support input with dynamic shapes, we need to expand the value to its dynamic dimension hence we need to create a sym_int_elements here.
```

File: torch_xla/csrc/tensor_methods.cpp [hunk 4]
```diff
@@ -1183,8 +1183,14 @@ void eye_out(XLATensorPtr& out, int64_t lines, int64_t cols) {
 }
 
 void fill_(XLATensorPtr& input, const at::Scalar& value) {
+  // Fill_ is implemented by expanding the pass in value to the same shape of
+  // input and replace input Tensor's IR with the expanded value. In order to
+  // support input with dynamic shapes, we need to expand the value to its
+  // dynamic dimension hence we need to create a sym_int_elements here.
+  SymIntElements sym_int_elements(input->GetIrValue(), input->shape());
```
Comment (JackCaoG) [inline torch_xla/csrc/tensor_methods.cpp:1190]: why couldn't we get the shape from IR value inside `SymIntElements` directly?

**Review by JackCaoG — APPROVED** [review]
Thanks!

**Pull request merged**